### PR TITLE
Fix STARK prover benchmarks

### PR DESCRIPTION
**Pull request by MegaRedHand**
# Fix benchmarks

## Description

This PR:
- fixes some benches that were panicking
- scope benches better, moving non-proof-generation stuff to the bench setup
- add benches with different blowup factors and number of queries
- add factorial Cairo programs and benches (they are a WIP though)
- add Cairo programs to IAI benches
- changed stark prover benchmark names for easier filtering
- add Cairo program compilation and trace generation to makefile and CI
- add flamegraph and benchmark filtering to make targets by `FILTER` env variable

## Type of change

- [X] Optimization

## Checklist
- [ ] Linked to Github Issue
- [ ] Unit tests added
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
- [ ] This change is an Optimization
  - [X] Benchmarks added/run

### CODE REVIEW

File: proving_system/stark/benches/criterion_stark.rs [hunk 1]
```diff
@@ -7,47 +17,73 @@ pub fn proof_benchmarks(c: &mut Criterion) {
...
+        group.throughput(Throughput::Elements(trace_length as u64));
+        group.bench_function(BenchmarkId::new("2 column Fibonacci", trace_length), |b| {
+            b.iter(|| black_box(prove(black_box(&trace), black_box(&fibonacci_air))));
+        });
+    }
 
-    group.bench_function("Quadratic AIR", |bench| {
-        bench.iter(functions::stark::prove_quadratic);
-    });
+    // TODO: this benchmarks panic
```
Comment (Oppen) [inline proving_system/stark/benches/criterion_stark.rs:36]: Do we know why?

File: Makefile [hunk 2]
```diff
@@ -13,15 +19,23 @@ docker-shell:
...
 
-benchmarks:
-	cargo criterion --bench all_benchmarks
+TEST_DIR=proving_system/stark/cairo_programs
+TEST_FILES:=$(wildcard $(TEST_DIR)/*.cairo)
+COMPILED_TESTS:=$(patsubst $(TEST_DIR)/%.cairo, $(TEST_DIR)/%.json, $(TEST_FILES))
+TEST_TRACES:=$(patsubst $(TEST_DIR)/%.cairo, $(TEST_DIR)/%.trace, $(TEST_FILES))
+
+$(TEST_DIR)/%.json: $(TEST_DIR)/%.cairo
+	cairo-compile --cairo_path="$(TEST_DIR)" $< --output $@
```
Comment (Oppen) [inline Makefile:28]: We're missing targets to actually run the program and produce the trace.

File: proving_system/stark/benches/criterion_stark.rs (hunk 1 above)
Comment (MegaRedHand) [reply proving_system/stark/benches/criterion_stark.rs]: I checked the integration tests and updated the parameters used in the benches to match those. Now they (hopefully) don't panic!

File: Makefile (hunk 2 above)
Comment (MegaRedHand) [reply Makefile:28]: Yes. I found it easier to produce them inside the benches (as part of the setup, NOT in the benchmarked code). It shouldn't add too much overhead to the total run time. What do you think?

Comment (Oppen) [reply Makefile:28]: In principle, it sounds good. But I'm worried it's not the exact same results. In fact I'm just looking at the `cairo_run` method in the project and it's not using proof mode. It's my understanding that it should run with that enabled.
Ideally I'd use the cairo-lang version for this.

Comment (MegaRedHand) [reply Makefile]: Done

**Review by schouhy — APPROVED** [review]

**Review by Oppen — APPROVED** [review]

File: proving_system/stark/benches/functions/stark.rs [hunk 3]
```diff
@@ -35,21 +46,26 @@ pub fn prove_fib(trace_length: usize) {
...
-    let result = prove(&trace, &fibonacci_air);
-    verify(&result, &fibonacci_air);
+    (trace, fibonacci_air)
 }
 
-pub fn prove_fib_2_cols() {
-    let trace_columns =
-        fibonacci_2_columns::fibonacci_trace_2_columns([FE::from(1), FE::from(1)], 16);
+#[allow(dead_code)]
+pub fn generate_fib_2_cols_proof_params(
```
Comment (MauroToscano) [inline proving_system/stark/benches/functions/stark.rs:53]: Shouldn't need a dead code here

File: proving_system/stark/benches/functions/stark.rs [hunk 4]
```diff
@@ -15,14 +20,20 @@ use lambdaworks_stark::air::example::{
 
 use crate::util::FE;
 
-pub fn prove_fib(trace_length: usize) {
+#[allow(dead_code)]
```
Comment (MauroToscano) [inline proving_system/stark/benches/functions/stark.rs:23]: Shouldn't need a dead code here. See if you need to compile for test or reduce visibility

**Review by MauroToscano — CHANGES_REQUESTED** [review]

**Review by Oppen — APPROVED** [review]

**Review by MauroToscano — APPROVED** [review]

**Pull request merged**